### ISSUE TITLE
train_OneLayerModel.py script and training samples errors

### ISSUE THREAD

**Issue by alshaboti** [Lab41/PoseidonML#50]
Hi, 
I am trying to create a model to be used in Poseidon #49. I captured traffic from raspberry pi and a laptop into two pcap files, then. 

1- I added "IOT" into config.json labels list. 
2- I added the two pcap files into pcap dir and edited label_assignments.json as below. 
```
root@e0e996977e0e:/app/NodeClassifier/pcap# ls
BusinessWS-60m-mylaptop.pcap  label_assignments.json
IOT-260m-ibmwatsoniot.pcap

root@e0e996977e0e:/app/NodeClassifier/pcap# cat label_assignments.json
{
    "IOT": "IOT",
    "BusinessWS": "Business workstation"
}
```

When I run train script, I got "ValueError: This solver needs samples of at least 2 classes in the data, but the data contains only one class: 0". 
What is wrong? I have already added two pcap files and labeled them in label_assignments.json. 

```
rs/ot@e0e996977e0e:/app/NodeClassifier# python train_OneLayerModel.py pcap/ models
Reading data
Reading pcap/IOT-260m-ibmwatsoniot.pcap as IOT
Reading pcap/BusinessWS-60m-mylaptop.pcap as Business workstation
Making data splits
Normalizing features
Doing feature selection
Traceback (most recent call last):
  File "train_OneLayerModel.py", line 28, in <module>
    model.train(data_dir)
  File "/app/NodeClassifier/utils/OneLayer.py", line 153, in train
    self.feature_list = select_features(X_normed, y_train)
  File "/app/NodeClassifier/utils/training_utils.py", line 101, in select_features
    selection_model.fit(X, y)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 112, in fit
    sample_fraction=self.sample_fraction, **params)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/memory.py", line 283, in __call__
    return self.func(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 54, in _resample_model
    for _ in range(n_resampling)):
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 758, in __call__
    while self.dispatch_one_batch(iterator):
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 608, in dispatch_one_batch
    self._dispatch(tasks)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 571, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/_parallel_backends.py", line 109, in apply_async
    result = ImmediateResult(func)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/_parallel_backends.py", line 326, in __init__
    self.results = batch()
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 131, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 131, in <listcomp>
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 377, in _randomized_logistic
    clf.fit(X, y)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/logistic.py", line 1186, in fit
    sample_weight=sample_weight)
  File "/usr/local/lib/python3.5/site-packages/sklearn/svm/base.py", line 875, in _fit_liblinear
    " class: %r" % classes_[0])
ValueError: This solver needs samples of at least 2 classes in the data, but the data contains only one class: 0
```

**Comment by CStephenson970** [comment]
Hi, 

It looks like you're doing everything right.  This seems like an issue with the feature selection step that can occur with small datasets.  Essentially a random sample is being drawn from the training data that only contains examples from one of your two classes.  Unfortunately I don't have a workaround ready at the moment, but I'm working on it.

**Comment by alshaboti** [comment]
Thanks for letting me know. 

The pcaps file are between 45 to 90 minutes of traffic.
1- What would be considered as a sufficient dataset? 6 hours or more? 
2- Is it necessary for the two classes samples (pcap files) to have the same number of packets (to be balanced)?
3- When it works with [Poseidon](https://github.com/CyberReboot/poseidon) does PoseidonML needs to be trained offline? if not then how does it work? Similar question I put in [Poseidon repo] (https://github.com/CyberReboot/poseidon/issues/473)

Basically, I started from this [Building a Software-Defined Network with Raspberry Pis and a Zodiac FX switch](https://blog.cyberreboot.org/building-a-software-defined-network-with-raspberry-pis-and-a-zodiac-fx-switch-97184032cdc1), but I didn't reach the point where Poseidon use PoseidonML to detect anomaly YET. 

Thank you for your reply.

**Comment by CStephenson970** [comment]
For question 1, the likely cause of this issue is that one of the pcaps doesn't contain much internal to internal IP traffic.  The PoseidonML plugins ignores traffic to public IP addresses.  As for the exact amount of internal to internal IP traffic that's sufficient, in our testing that seems to be dataset specific, and might need some experimentation with to get good results.  Several hours or more (~5) would be ideal.

For question 2, it is not necessary that the pcap files contain the same number of packets, but you may get poor performance if you have large imbalances (ex: 30 minutes of collection time on one type vs 10 hours on another) 

For question 3, as of right now, it is necessary to train the ML models offline, although this should change in the near future.

**Comment by alshaboti** [comment]
Thank you so much, 
That's answers all my questions. 

I will try with another pcap files with internal to internal for more than 5hours, and get back to you.

**Comment by gregs5** [comment]
@alshaboti , if it helps any, I used an IoT data set from a New South Wales research effort to do training. I can say that after I extracted single nodes from a pcap containing MANY nodes, the training process did work.  However, the approach that has been taken with PoseidonML at THIS STAGE is probably better suited for more "enterprise-like" nodes vs. IoT ones.  (Think: a lot more "north-south" traffic with IoT over "east-west" like traffic.)  

I don't know what your data sources are, just thought I'd pass that on in case it is useful!

**Comment by alshaboti** [comment]
Thanks @gregs5 
Yes, that will help. My target is to use Poseidon with smart-home IoT. But for now, I am trying to run it with any dataset regardless what the performance is. 

Could you please share that dataset with me?

**Comment by gregs5** [comment]
@alshaboti : absolutely.  I found it here: [http://149.171.189.1/](http://149.171.189.1/)

Keep in mind that you'll need to extract and label individual nodes from the bigger pcap file.  I did this using (roughly):

`tcpdump ether host <mac addr> -r <pcap input file> -w <pcap output file>`

...but I'm sure there are other ways of doing it. I'm really curious what results you get from your own IoT data; I found that our models don't work great for this type of data, but it's still worth experimenting with, IMO.  Would be thrilled to hear what you find!

**Comment by gregs5** [comment]
And the paper: http://www2.ee.unsw.edu.au/~vijay/pubs/conf/17infocom.pdf

**Comment by alshaboti** [comment]
Thank you very much,  
that's  what I was looking for. 
Surely, I will share my findings.

**Comment by alshaboti** [comment]
I extract traffic for 3 IoT and train the onelayer ML , then save the model into models/model.pickel. 
when I use it with Poseidon, I got this error. 

```
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]: DEBUG:__main__:Could not get address info beacuse list index out of range
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]: DEBUG:__main__:Defaulting to inferring IP address from
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]: DEBUG:__main__:Loaded model from models/model.pickle
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]: Traceback (most recent call last):
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]:   File "eval_OneLayer.py", line 362, in <module>
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]:     mean=False
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]:   File "/NodeClassifier/utils/OneLayer.py", line 227, in get_representation
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]:     source_ip=source_ip,
2018-01-07T15:47:46+00:00 172.17.0.1 plugin[1201]: ValueError: not enough values to unpack (expected 4, got 3)
```

There is extra comma in this line
```
https://github.com/Lab41/PoseidonML/blob/master/NodeClassifier/utils/OneLayer.py#L227
```

**Comment by alshaboti** [comment]
This error has been solved by following the naming pattern for pcap file like the same [here](https://github.com/Lab41/PoseidonML/tree/master/NodeClassifier/data)

**Comment by cglewis** [comment]
@alshaboti is this ok to close then?

**Comment by alshaboti** [comment]
Yes, I will close it. 
Thank you,

**Comment by jaiken06** [comment]
@CStephenson970 

Was a work around ever introduced for this?  I am getting a similar error using my one hour long pcaps, is this because they still need to be a good few hours longer? Thanks

**Comment by cglewis** [comment]
@jaiken06 which error?  Can you describe the kind of traffic you're using? Is it internal to internal? public only? ipv6? tcp/udp? Depending on the traffic type it might be getting ignored.

**Comment by jaiken06** [comment]
It is just under one hour of web traffic that I recorded of myself just surfing the internet.  I then forged the source and destination IP addresses of the packets so I could send it from one Pi to another through my Zodiac.

It is just HTTP, TCP, DNS etc.  The error is as follows:

```
Reading data
Reading ./Training/WebTraffic-test.pcap as WebTraffic
Making data splits
Normalizing features
/app/NodeClassifier/utils/OneLayer.py:148: RuntimeWarning: Mean of empty slice.
 self.means = X_train.mean(axis=0)
/usr/local/lib/python3.5/site-packages/numpy/core/_methods.py:73: RuntimeWarning: invalid value encountered in true_divide
 ret, rcount, out=ret, casting='unsafe', subok=False)
/usr/local/lib/python3.5/site-packages/numpy/core/_methods.py:135: RuntimeWarning: Degrees of freedom <= 0 for slice
 keepdims=keepdims)
/usr/local/lib/python3.5/site-packages/numpy/core/_methods.py:105: RuntimeWarning: invalid value encountered in true_divide
 arrmean, rcount, out=arrmean, casting='unsafe', subok=False)
/usr/local/lib/python3.5/site-packages/numpy/core/_methods.py:125: RuntimeWarning: invalid value encountered in true_divide
 ret, rcount, out=ret, casting='unsafe', subok=False)
/app/NodeClassifier/utils/OneLayer.py:151: RuntimeWarning: invalid value encountered in less_equal
 zero_stds = self.stds <= 1
Doing feature selection
Traceback (most recent call last):
 File "train_OneLayer.py", line 31, in <module>
   model.train(data_dir)
 File "/app/NodeClassifier/utils/OneLayer.py", line 159, in train
   self.feature_list = select_features(X_normed, y_train)
 File "/app/NodeClassifier/utils/training_utils.py", line 101, in select_features
   selection_model.fit(X, y)
 File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 93, in fit
   ensure_min_samples=2, estimator=self)
 File "/usr/local/lib/python3.5/site-packages/sklearn/utils/validation.py", line 521, in check_X_y
   ensure_min_features, warn_on_dtype, estimator)
 File "/usr/local/lib/python3.5/site-packages/sklearn/utils/validation.py", line 416, in check_array
   context))
ValueError: Found array with 0 sample(s) (shape=(0, 4104)) while a minimum of 2 is required by RandomizedLogisticRegression.
```


I downloaded some pcaps from the link suggested above and they weren't getting this error (presumably because they were much larger).

**Comment by cglewis** [comment]
So I believe the default expects tcp/udp sessions that are at least 2 minutes long. and then it expects some minimum number of sessions, I believe - not sure on that. the 2 minutes long comes from `session threshold` defined here: https://github.com/CyberReboot/PoseidonML/blob/master/NodeClassifier/config.json

I wouldn't recommend tweaking that, as I don't know how that will change the results, but you could try setting it to 1 to see if the errors go away, which would indicate that the sessions in your pcap aren't long enough.

**Comment by jaiken06** [comment]
I have changed that threshold to 1 and am using a 600mb pcap from the link mentioned in this issue and am now receiving the following error:

```
Reading data
Reading ./Training/WebTraffic-test.pcap as WebTraffic
Making data splits
Normalizing features
Doing feature selection
Traceback (most recent call last):
  File "train_OneLayer.py", line 31, in <module>
    model.train(data_dir)
  File "/app/NodeClassifier/utils/OneLayer.py", line 159, in train
    self.feature_list = select_features(X_normed, y_train)
  File "/app/NodeClassifier/utils/training_utils.py", line 101, in select_features
    selection_model.fit(X, y)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 112, in fit
    sample_fraction=self.sample_fraction, **params)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/memory.py", line 283, in __call__
    return self.func(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 54, in _resample_model
    for _ in range(n_resampling)):
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 758, in __call__
    while self.dispatch_one_batch(iterator):
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 608, in dispatch_one_batch
    self._dispatch(tasks)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 571, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/_parallel_backends.py", line 109, in apply_async
    result = ImmediateResult(func)
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/_parallel_backends.py", line 326, in __init__
    self.results = batch()
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 131, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py", line 131, in <listcomp>
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/randomized_l1.py", line 377, in _randomized_logistic
    clf.fit(X, y)
  File "/usr/local/lib/python3.5/site-packages/sklearn/linear_model/logistic.py", line 1186, in fit
    sample_weight=sample_weight)
  File "/usr/local/lib/python3.5/site-packages/sklearn/svm/base.py", line 875, in _fit_liblinear
    " class: %r" % classes_[0])
ValueError: This solver needs samples of at least 2 classes in the data, but the data contains only one class: 0
```

**Comment by cglewis** [comment]
It might be worth throwing your pcap into wireshark and looking at the sessions to see if anything looks strange?  How many classes are in your `label_assignments.json`?

**Comment by jaiken06** [comment]
I only have one class in `label_assignments.json` currently just as I was trying to see if I could get a train working, could this be a problem, does it require at least 2 labels?

**Comment by cglewis** [comment]
yes, at least 2, the more the better.

**Comment by jaiken06** [comment]
Unfortunately I am still getting the same error:

`ValueError: This solver needs samples of at least 2 classes in the data, but the data contains only one class: 0`

I did notice a comment above stated:

``` if it helps any, I used an IoT data set from a New South Wales research effort to do training. I can say that after I extracted single nodes from a pcap containing MANY nodes, the training process did work.
```
...
```
Keep in mind that you'll need to extract and label individual nodes from the bigger pcap file. I did this using (roughly):

tcpdump ether host <mac addr> -r <pcap input file> -w <pcap output file>
```

What is the reason for this? So do I just tcpdump out the traffic for specific mac addresses and record that using wireshark?  Why does it only work if it is traffic of a certain node?

**Comment by jaiken06** [comment]
Adding in more labels and following the advice further up this thread about getting the traffic for certain Nodes/Mac addresses and using those as labels, it no longer says "This solver needs samples of at least 2 classes in the data".

The error is the following:

```
Reading data
Reading ./Training/IOT2-test.pcap as IOT2
Reading ./Training/IOT1-test.pcap as IOT1
Reading ./Training/IOT3-test.pcap as IOT3
Making data splits
Normalizing features
Doing feature selection
[0, 53, 67, 80, 123, 443, 465, 1024, 1077, 1091, 1092, 1104, 1147, 1467, 4096, 4097, 4098, 4099]
Traceback (most recent call last):
  File "train_OneLayer.py", line 31, in <module>
    model.train(data_dir)
  File "/app/NodeClassifier/utils/OneLayer.py", line 172, in train
    X_aug, y_aug = self._augment_data(X_normed, y_train)
  File "/app/NodeClassifier/utils/OneLayer.py", line 55, in _augment_data
    y_permed = [self.labels.index("Unknown")]*X_permed.shape[0]
ValueError: 'Unknown' is not in list
```

**Comment by cglewis** [comment]
@jaiken06 glad it's working now.  I didn't realize it was required to have `unknown` in your config, sorry about that.
